src/lib/core/CHIPTLVReader.cpp

```diff
@@ -125,81 +126,125 @@ CHIP_ERROR TLVReader::Get(bool & v)
     return CHIP_NO_ERROR;
 }
 
+template<typename int_type>
+static inline bool TLVTypeValid(bool tlv_isSigned, uint64_t val64)
+{
+    return !(tlv_isSigned ^ std::is_signed<int_type>::value) &&
+	    (tlv_isSigned && val64 >> 63) ? CanCastTo<int_type>(-val64) : CanCastTo<int_type>(val64);
+}
+
 CHIP_ERROR TLVReader::Get(int8_t & v)
 {
     uint64_t v64   = 0;
-    CHIP_ERROR err = Get(v64);
-    v              = CastToSigned(static_cast<uint8_t>(v64));
+    bool tlv_isSigned = false;
+
+    CHIP_ERROR err = Get(v64, tlv_isSigned);
+    ReturnErrorCodeIf(!TLVTypeValid<int8_t>(tlv_isSigned, v64), CHIP_ERROR_WRONG_TLV_TYPE);
+    v = CastToSigned(static_cast<uint8_t>(v64));
```

Comment:
This might be clearer as `v = static_cast<int8_t>(CastToSigned(v64));` just because `TLVTypeValid` ended up checking that `CanCastTo<int8_t>(CastToSigned(v64))` is true, so this would be obviously symmetric with that.  But I think I've convinced myself that this produces the same result.  Again, we should add tests for various possible edge cases.  Again, applies to the other functions.